File: src/strategies/retry.ts
import type { RetryConfig } from "../core/types.js";
import { IdempotencyLevel } from "../core/types.js";
import type { IdempotencyResolver } from "./idempotency.js";

export class RetryStrategy {
  private config: Required<RetryConfig>;

  constructor(config: Partial<RetryConfig> = {}, _idempotencyResolver: IdempotencyResolver) {
    this.config = {
      maxRetries: config.maxRetries ?? 0,
      baseDelay: config.baseDelay ?? 1000,
      maxDelay: config.maxDelay ?? 30000,
      jitter: config.jitter ?? true,
    };
  }

  async execute<T>(
    fn: () => Promise<T>,
    idempotencyLevel: IdempotencyLevel,
    hasIdempotencyKey: boolean,
    attempt = 0,
  ): Promise<T> {
    try {
      return await fn();
    } catch (error) {
      if (attempt >= this.config.maxRetries) {
        throw error;
      }

      if (!this.isExplicitlyRetryable(error)) {
        throw error;
      }

      if (!this.isIdempotencyProven(idempotencyLevel, hasIdempotencyKey)) {
        throw error;
      }

      const delay = this.calculateDelay(attempt);

      await this.sleep(delay);

      return this.execute(fn, idempotencyLevel, hasIdempotencyKey, attempt + 1);
    }
  }

  private isExplicitlyRetryable(error: unknown): boolean {
    if (error && typeof error === "object" && "retryable" in error) {
      return (error as { retryable: boolean }).retryable === true;
    }

    return false;
  }

  private isIdempotencyProven(
    idempotencyLevel: IdempotencyLevel,
    hasIdempotencyKey: boolean,
  ): boolean {
    switch (idempotencyLevel) {
      case IdempotencyLevel.SAFE:
      case IdempotencyLevel.IDEMPOTENT:
        return true;
      case IdempotencyLevel.CONDITIONAL:
        return hasIdempotencyKey;
      case IdempotencyLevel.UNSAFE:
        return false;
      default:
        return false;
    }
  }

  private calculateDelay(attempt: number): number {
    let delay = this.config.baseDelay * Math.pow(2, attempt);

    if (this.config.jitter) {
      const jitter = Math.random() * 1000;
      delay += jitter;
    }

    return Math.min(delay, this.config.maxDelay);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: clients/python/meridian/strategies/retry.py
"""Port of src/strategies/retry.ts."""

from __future__ import annotations

import asyncio
import random
from typing import Awaitable, Callable, TypeVar

from ..contract import IdempotencyLevel

T = TypeVar("T")


class RetryStrategy:
    def __init__(self, config: dict | None = None) -> None:
        config = config or {}
        self.max_retries: int = config.get("max_retries", 0)
        self.base_delay: float = config.get("base_delay", 1000)
        self.max_delay: float = config.get("max_delay", 30000)
        self.jitter: bool = config.get("jitter", True)

    async def execute(
        self,
        fn: Callable[[], Awaitable[T]],
        idempotency_level: IdempotencyLevel,
        has_idempotency_key: bool,
        attempt: int = 0,
    ) -> T:
        try:
            return await fn()
        except Exception as error:  # noqa: BLE001 — mirror the TS catch-all
            if attempt >= self.max_retries:
                raise
            if not self._is_explicitly_retryable(error):
                raise
            if not self._is_idempotency_proven(idempotency_level, has_idempotency_key):
                raise

            delay_ms = self._calculate_delay(attempt)
            await asyncio.sleep(delay_ms / 1000)
            return await self.execute(
                fn, idempotency_level, has_idempotency_key, attempt + 1
            )

    @staticmethod
    def _is_explicitly_retryable(error: Exception) -> bool:
        return getattr(error, "retryable", None) is True

    @staticmethod
    def _is_idempotency_proven(
        idempotency_level: IdempotencyLevel, has_idempotency_key: bool
    ) -> bool:
        if idempotency_level in (IdempotencyLevel.SAFE, IdempotencyLevel.IDEMPOTENT):
            return True
        if idempotency_level == IdempotencyLevel.CONDITIONAL:
            return has_idempotency_key
        return False

    def _calculate_delay(self, attempt: int) -> float:
        delay = self.base_delay * (2**attempt)
        if self.jitter:
            delay += random.random() * 1000
        return min(delay, self.max_delay)
